qiskit/dagcircuit/dagcircuit.py

```diff
@@ -319,6 +319,113 @@ def _add_wire(self, wire):
         else:
             raise DAGCircuitError(f"duplicate wire {wire}")
 
+    def remove_clbits(self, *clbits):
+        """
+        Remove classical bits from the circuit. Only the bits that are idle and not
+        referenced by any classical registers will be removed. Other bits are ignored.
+
+        Raises:
+            DAGCircuitError: a clbit is not a Clbit.
+
+        Returns:
+            set(Clbit): The set of Clbits that were removed.
+        """
+        if any(not isinstance(clbit, Clbit) for clbit in clbits):
+            raise DAGCircuitError("not a Clbit instance.")
+
+        # ignore clbits not in circuit
+        clbits = set(clbits).intersection(self.clbits)
+
+        # ignore clbits referenced by registers
+        register_clbits = {bit for creg in self.cregs.values() for bit in creg}
+        clbits -= register_clbits
+
+        # ignore clbits that are busy
+        busy_clbits = {bit for bit in clbits if not self._is_wire_idle(bit)}
+        clbits -= busy_clbits
+
+        for clbit in clbits:
+            self._remove_idle_wire(clbit)
+            self.clbits.remove(clbit)
+
+        return clbits
+
+    def remove_cregs(self, *cregs):
+        """
+        Remove classical registers from the circuit,
+        removing idle bits that are no longer referenced by any register
+        as a result of this operation.
+
+        Raises:
+            DAGCircuitError: a creg is not a ClassicalRegister.
+        Returns:
+            set(Clbit): the bits removed from the circuit as a result of this operation.
+        """
+        if any(not isinstance(creg, ClassicalRegister) for creg in cregs):
+            raise DAGCircuitError("not a ClassicalRegister instance.")
+
+        # Ignore any cregs not in circuit.
+        cregs = set(cregs).intersection(self.cregs.values())
+
+        if not cregs:
+            return set()
+
+        cregs_to_remove = set()
+        clbits_to_remove = set()
+        clbits_to_keep = set()
+        for creg in self.cregs.values():
+            clbits = set(creg)
+            if creg in cregs:
+                cregs_to_remove.add(creg)
+                clbits_to_remove.update(clbits)
+            else:
+                clbits_to_keep.update(clbits)
+
+        clbits_to_remove.difference_update(clbits_to_keep)
+
+        for clbit in clbits_to_remove:
+            if self._is_wire_idle(clbit):
+                self._remove_idle_wire(clbit)
+                self.clbits.remove(clbit)
+            else:
+                # TODO: this would probably be better as a debug log statement
+                warnings.warn("clbit %s is not idle but is no longer referenced by a register")
```

Comment:
I don't think this lines up with the current state of Terra, necessarily - a bit can be in more than one register already, so I can do something like this:
```python
In [1]: from qiskit.circuit import QuantumCircuit, Qubit, QuantumRegister
   ...: bits = [Qubit(), Qubit()]
   ...: regs = [QuantumRegister(bits=bits) for _ in [None]*5]
   ...: qc = QuantumCircuit(*regs)
```
without error.  If I only passed one of those registers to `remove_cregs`, I'd see this warning, even though the bits _were_ referenced.

Either way, I don't think this is something we need to warn about - registers are (slowly) becoming just aliases for groups of bits; it's already not necessary for a given bit to be in a register in a circuit.

Reply:
Your code example is supposed to have `ClassicalRegister` in place of `QuantumRegister`, yeah?

> I'd see this warning, even though the bits were referenced.

This warning shouldn't be emitted in this case, unless I have a bug here that you're seeing that I don't! In this loop, we're (supposed) to be going over only the `clbits` that are candidates for removal. That is, all bits that appear in a register that will be removed, but do _not_ appear in a register that will be kept. If we are removing only 1 of the 5 registers above, we won't try to remove any of its bits, and so we won't see the warning.

Though I do agree with you: it shouldn't be a warning when a bit is left without a reference. I thought it might be nice to log it at least, but wasn't sure how to do that in this context (I don't see any traces in this class). If that's not useful, I'm happy to remove this `else` block completely.

Reply:
Removed the `else` block for now :).